packages/squiggle-lang/src/rescript/GenericDist/GenericDist.res

```diff
@@ -0,0 +1,253 @@
+//TODO: multimodal, add interface, test somehow, track performance, refactor sampleSet, refactor ASTEvaluator.res.
+type t = GenericDist_Types.genericDist
+type error = GenericDist_Types.error
+type toPointSetFn = t => result<PointSetTypes.pointSetDist, error>
+type toSampleSetFn = t => result<array<float>, error>
+type scaleMultiplyFn = (t, float) => result<t, error>
+type pointwiseAddFn = (t, t) => result<t, error>
+
+let sampleN = (t: t, n) =>
+  switch t {
+  | #PointSet(r) => Ok(PointSetDist.sampleNRendered(n, r))
+  | #Symbolic(r) => Ok(SymbolicDist.T.sampleN(n, r))
+  | #SampleSet(_) => Error(GenericDist_Types.NotYetImplemented)
+  }
+
+let fromFloat = (f: float) => #Symbolic(SymbolicDist.Float.make(f))
+
+let toString = (t: t) =>
+  switch t {
+  | #PointSet(_) => "Point Set Distribution"
+  | #Symbolic(r) => SymbolicDist.T.toString(r)
+  | #SampleSet(_) => "Sample Set Distribution"
+  }
+
+let normalize = (t: t) =>
+  switch t {
+  | #PointSet(r) => #PointSet(PointSetDist.T.normalize(r))
+  | #Symbolic(_) => t
+  | #SampleSet(_) => t
+  }
+
+let operationToFloat = (t, toPointSet: toPointSetFn, fnName) => {
+  let symbolicSolution = switch t {
+  | #Symbolic(r) =>
+    switch SymbolicDist.T.operate(fnName, r) {
+    | Ok(f) => Some(f)
+    | _ => None
+    }
+  | _ => None
+  }
+
+  switch symbolicSolution {
+  | Some(r) => Ok(r)
+  | None => toPointSet(t)->E.R2.fmap(PointSetDist.operate(fnName))
+  }
+}
+
+//TODO: Refactor this bit.
+let defaultSamplingInputs: SamplingInputs.samplingInputs = {
+  sampleCount: 10000,
+  outputXYPoints: 10000,
+  pointSetDistLength: 1000,
+  kernelWidth: None,
+}
+
+//Todo: If it's a pointSet, but the xyPointLenght is different from what it has, it should change.
+// This is tricky because the case of discrete distributions.
+let toPointSet = (t, xyPointLength): result<PointSetTypes.pointSetDist, error> => {
+  switch t {
+  | #PointSet(pointSet) => Ok(pointSet)
+  | #Symbolic(r) => Ok(SymbolicDist.T.toPointSetDist(xyPointLength, r))
+  | #SampleSet(r) => {
+      let response = SampleSet.toPointSetDist(
+        ~samples=r,
+        ~samplingInputs=defaultSamplingInputs,
+        (),
+      ).pointSetDist
+      switch response {
+      | Some(r) => Ok(r)
+      | None => Error(Other("Converting sampleSet to pointSet failed"))
+      }
+    }
+  }
+}
+
+module Truncate = {
+  let trySymbolicSimplification = (leftCutoff, rightCutoff, t): option<t> =>
+    switch (leftCutoff, rightCutoff, t) {
+    | (None, None, _) => None
+    | (lc, rc, #Symbolic(#Uniform(u))) if lc < rc =>
+      Some(#Symbolic(#Uniform(SymbolicDist.Uniform.truncate(lc, rc, u))))
+    | _ => None
+    }
+
+  let run = (
+    t: t,
+    toPointSet: toPointSetFn,
+    leftCutoff: option<float>,
```

Comment:
(question): Do we prefer `~leftCutoff = ?`?

Reply:
Yea, it's better, I just didn't get to it here. Not expecting many people to use GenericDist calls directly, but can set that up.